### ISSUE TITLE
Chapter 3. Page 100. ZeroDivisionError: integer division or modulo by zero

### ISSUE THREAD

**Issue by qiaozhi9092** [ageron/handson-ml#366]
Hi ageron,
When I was trying this piece of code:
<img width="612" alt="screen shot 2019-03-03 at 12 23 29 am" src="https://user-images.githubusercontent.com/10824922/53692798-da3fcf80-3d4a-11e9-8f8a-5083cd50c014.png">
the error message given is:
<img width="618" alt="screen shot 2019-03-03 at 12 23 41 am" src="https://user-images.githubusercontent.com/10824922/53692801-e88deb80-3d4a-11e9-8cef-a488e57da16c.png">
I did defined the function plot_digits() by copying from the notebook. 
Can you place help me on that? Thank you so much!

**Comment by ageron** [comment]
Hi @qiaozhi9092 ,

To debug, first look at the error message, and where it occurs: the message says there is a division by zero error at the line that computes `n_rows = (len(instances) - 1) // images_per_row + 1`.
As you can see, the only division on this line is a division by `images_per_row`, so it means it must be equal to 0. So next, let's look at where it is defined. Going up a couple line you find that it is defined as: `images_per_row = min(len(instances), images_per_row)`. Since we know that the result is 0, it means that either `len(instances)` is equal to 0, or `images_per_row` is equal to 0 (or both). Since your code calls `plot_digits(X_aa[:25], images_per_row=5)`, we see that it must be that `len(instances)` is 0.
In other words, `X_aa[:25]` is empty. Next, let's look at where this one is defined:

```python
X_aa = X_train[(y_train == cl_a) & (y_train_pred == cl_a)]
```

It seems that there is no instance in the training set that corresponds to these two conditions: `y_train == cl_a` and `y_train_pred == cl_a`.

One possible reason might be if you loaded the training set using `fetch_openml()`: it returns labels as strings, not integers. This is why in the notebook I make sure to convert the labels to integers just after loading the dataset:

```python
mnist.target = mnist.target.astype(np.int8) # fetch_openml() returns targets as strings
```

Hope this helps.

**Comment by qiaozhi9092** [comment]
Hi @ageron,

Thank you so much for your prompt reply. You made a very clear instruction of how to debug in this scenario. I followed your procedure and was able to find that the (y_train_pred == cl_a) was empty so it trace back to my y_train_pred was incorrectly defined (miss spilled, stupid me). 

Thank you again for your help, I am enjoying every little bit of your book!

Best,
